fenzo-core/src/main/java/com/netflix/fenzo/queues/tiered/SortedBuckets.java

```diff
@@ -81,13 +84,54 @@ QueueBucket remove(String bucketName) {
             remPos = findWalkingLeft(buckets, index, bucketName, bucket.getDominantUsageShare());
         if (remPos < 0)
             remPos = findWalkingRight(buckets, index, bucketName, bucket.getDominantUsageShare());
-        if (remPos < 0)
-            throw new IllegalStateException("Unexpected: bucket with name=" + bucketName + " not found to remove");
-        buckets.remove(remPos);
+        if (remPos < 0) {
+            logger.error("Unexpected: bucket with name=" + bucketName + " not found to remove, traversing " +
+                    buckets.size() + " buckets to remove it");
+            logger.warn("Invalid sorted buckets list: " + getBucketsListString());
+            removeBucketAndResort(bucketName);
+        }
+        else
+            buckets.remove(remPos);
         bucketMap.remove(bucketName);
         return bucket;
     }
 
+    private void removeBucketAndResort(String bucketName) {
+        // workaround the problem: linear traversal of the list to remove the bucket and re-sort if needed
+        if (!buckets.isEmpty()) {
+            final Iterator<QueueBucket> iterator = buckets.iterator();
+            QueueBucket prev = null;
+            boolean isSorted = true;
+            while (iterator.hasNext()) {
+                QueueBucket b = iterator.next();
+                if (b.getName().equals(bucketName)) {
+                    iterator.remove();
+                    if (!isSorted)
+                        break; // no need to check further
+                } else {
+                    if (prev != null) {
+                        final int compare = comparator.compare(prev, b);
+                        isSorted = compare <= 0;
```

Comment:
There should not be any duplicates since there is the bucketMap.containsKey() which is checked before adding. 
I can add it as an additional error to check on, in case that there is a different bug causing that.